File: flower/src/main/java/com/kharitonov/flower/services/AuthService.java
package com.kharitonov.flower.services;

import com.kharitonov.flower.contracts.JwtRequest;
import com.kharitonov.flower.contracts.JwtResponse;
import com.kharitonov.flower.entities.UserEntity;
import com.kharitonov.flower.exceptions.AuthException;
import com.kharitonov.flower.security.JwtAuthentication;
import com.kharitonov.flower.security.JwtProvider;
import io.jsonwebtoken.Claims;
import lombok.NonNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;

@Service
public class AuthService {

    private final UserService userService;

    //Будет redis...
    private final Map<String, String> refreshStorage = new HashMap<>();
    private final JwtProvider jwtProvider;

    @Autowired
    public AuthService(UserService userService, JwtProvider jwtProvider) {
        this.userService = userService;
        this.jwtProvider = jwtProvider;
    }



    public JwtResponse login(@NonNull JwtRequest authRequest) {
        final UserEntity user = userService.getByLogin(authRequest.getLogin())
                .orElseThrow(() -> new AuthException("User not found"));
        if (user.getPassword().equals(authRequest.getPassword())) {
            final String accessToken = jwtProvider.generateAccessToken(user);
            final String refreshToken = jwtProvider.generateRefreshToken(user);
            refreshStorage.put(user.getLogin(), refreshToken);
            return new JwtResponse(accessToken, refreshToken);
        } else {
            throw new AuthException("Wrong password");
        }
    }

    public JwtResponse getAccessToken(@NonNull String refreshToken) {
        if (jwtProvider.validateRefreshToken(refreshToken)) {
            final Claims claims = jwtProvider.getRefreshClaims(refreshToken);
            final String login = claims.getSubject();
            final String saveRefreshToken = refreshStorage.get(login);
            if (saveRefreshToken != null && saveRefreshToken.equals(refreshToken)) {
                final UserEntity user = userService.getByLogin(login)
                        .orElseThrow(() -> new AuthException("User not found"));
                final String accessToken = jwtProvider.generateAccessToken(user);
                return new JwtResponse(accessToken, null);
            }
        }
        return new JwtResponse(null, null);
    }

    public JwtResponse refresh(@NonNull String refreshToken) {
        if (jwtProvider.validateRefreshToken(refreshToken)) {
            final Claims claims = jwtProvider.getRefreshClaims(refreshToken);
            final String login = claims.getSubject();
            final String saveRefreshToken = refreshStorage.get(login);
            if (saveRefreshToken != null && saveRefreshToken.equals(refreshToken)) {
                final UserEntity user = userService.getByLogin(login)
                        .orElseThrow(() -> new AuthException("User not found"));
                final String accessToken = jwtProvider.generateAccessToken(user);
                final String newRefreshToken = jwtProvider.generateRefreshToken(user);
                refreshStorage.put(user.getLogin(), newRefreshToken);
                return new JwtResponse(accessToken, newRefreshToken);
            }
        }
        throw new AuthException("Invalid JWT token");
    }

    public JwtAuthentication getAuthInfo() {
        return (JwtAuthentication) SecurityContextHolder.getContext().getAuthentication();
    }

}

File: src/main/java/by/hghgrtut/shop/services/AuthService.kt
package by.hghgrtut.shop.services

import by.hghgrtut.shop.contracts.JwtRequest
import by.hghgrtut.shop.contracts.JwtResponse
import by.hghgrtut.shop.entities.UserEntity
import by.hghgrtut.shop.exceptions.AuthException
import by.hghgrtut.shop.security.JwtAuthentication
import by.hghgrtut.shop.security.JwtProvider
import io.jsonwebtoken.Claims
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import java.util.HashMap
import java.util.Map

@Service
class AuthService @Autowired constructor(private val userService: UserService, private val jwtProvider: JwtProvider) {

    val authInfo: JwtAuthentication get() = SecurityContextHolder.getContext().getAuthentication() as JwtAuthentication

    private val refreshStorage: Map<String, String> = HashMap()

    fun login(authRequest: JwtRequest): JwtResponse {
        val user: UserEntity = userService.getByLogin(authRequest.getLogin()).orThrowUserNotFoundException()
        return if (user.getPassword().equals(authRequest.getPassword())) {
            val refreshToken: String = jwtProvider.generateRefreshToken(user)
            refreshStorage.put(user.getLogin(), refreshToken)
            JwtResponse(jwtProvider.generateAccessToken(user), refreshToken)
        } else { throw AuthException("Wrong password") }
    }

    fun getAccessToken(refreshToken: String): JwtResponse {
        var accessToken: String? = null
        if (jwtProvider.validateRefreshToken(refreshToken)) {
            val login: String = jwtProvider.getRefreshClaims(refreshToken).getSubject()
            refreshStorage.get(login)?.let {
                if (it.equals(refreshToken))
                    accessToken =
                        jwtProvider.generateAccessToken(userService.getByLogin(login).orThrowUserNotFoundException())
            }
        }
        return JwtResponse(accessToken, null)
    }

    fun refresh(refreshToken: String): JwtResponse {
        if (jwtProvider.validateRefreshToken(refreshToken)) {
            val login: String = jwtProvider.getRefreshClaims(refreshToken).getSubject()
            refreshStorage.get(login)?.let {
                if (it.equals(refreshToken)) {
                    val user: UserEntity = userService.getByLogin(login).orThrowUserNotFoundException()
                    val newRefreshToken: String = jwtProvider.generateRefreshToken(user)
                    refreshStorage.put(user.getLogin(), newRefreshToken)
                    return JwtResponse(jwtProvider.generateAccessToken(user), newRefreshToken)
                }
            }
        }
        throw AuthException("Invalid JWT token")
    }

    private fun String.orThrowUserNotFoundException() = orElseThrow { AuthException("User not found") }
}